src/main/java/au/org/ala/names/index/CSVNameSource.java

```diff
@@ -0,0 +1,240 @@
+package au.org.ala.names.index;
+
+import au.org.ala.vocab.ALATerm;
+import au.com.bytecode.opencsv.CSVReader;
+import au.org.ala.names.model.RankType;
+import au.org.ala.names.model.TaxonomicType;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.StringField;
+import org.gbif.api.model.registry.Citation;
+import org.gbif.api.model.registry.Contact;
+import org.gbif.api.vocabulary.Country;
+import org.gbif.api.vocabulary.NomenclaturalCode;
+import org.gbif.api.vocabulary.NomenclaturalStatus;
+import org.gbif.dwc.terms.DwcTerm;
+import org.gbif.dwc.terms.Term;
+import org.gbif.dwc.terms.TermFactory;
+
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * A source of names from a CSV file.
+ * <p>
+ * The first row must contain
+ * </p>
+ *
+ * @author Doug Palmer &lt;Doug.Palmer@csiro.au&gt;
+ * @copyright Copyright (c) 2016 CSIRO
+ */
+public class CSVNameSource extends NameSource {
+    private String name;
+    private CSVReader reader;
+    private List<Term> terms;
+    private Map<Term, Integer> termLocations;
+
+    /**
+     * Open a CSV source.
+     *
+     * @param reader The
+     */
+    public CSVNameSource(Reader reader) throws IOException {
+        this.name = "Reader " + reader.toString();
+        this.reader = new CSVReader(reader);
+        this.collectColumns();
+    }
+
+    /**
+     * Open a file as a CSV name source.
+     *
+     * @param file The source file
+     * @param encoding The source encoding
+     */
+    public CSVNameSource(File file, String encoding) throws IOException {
+        this(new InputStreamReader(new FileInputStream(file), encoding));
+        this.name = file.getCanonicalPath();
+    }
+
+    protected void collectColumns() throws IOException {
+        TermFactory factory = TermFactory.instance();
+        int index = 0;
+        String[] header = reader.readNext();
+
+        if (header == null || header.length == 0)
+            throw new IndexBuilderException("No header in CSV file");
+        this.termLocations = new HashMap<>(header.length);
+        this.terms = new ArrayList<>(header.length);
+        for (String heading: header) {
+            heading = heading.trim();
+            Term term = factory.findTerm(heading);
+            if (term == null)
+                term = ALATerm.valueOf(heading);
+            if (term != null)
+                this.termLocations.put(term, index);
+            this.terms.add(term);
```

Comment:
"This should never happen" Basically trying to keep the term positions consistent, even if the term lookup failed, which is possible because term lookup may not succeed.

Reply:
Also logging here for missing terms